### PR TITLE
Refactor leaf node handling in smt2_conv

### PR DESCRIPTION
**Pull request by NlightNFotis**
This is for internal reviewing of the refactoring of the smt2_conv leaf nodes, in anticipation of integration with smt-switch.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by kroening** [comment]
Please remember the promise that CBMC needs to be compilable without SMT-Switch!

**Comment by NlightNFotis** [comment]
It will be. This PR will just host a refactoring of smt2-conv.cpp - without anything to do with smt-switch. That will come at a later PR.

File: src/solvers/smt2/smt2_conv.cpp [hunk 1]
```diff
@@ -1257,7 +1257,13 @@ void smt2_convt::convert_expr(const exprt &expr)
   }
   else if(expr.id()==ID_mod)
   {
-    convert_mod(to_mod_expr(expr));
+    // TODO: change to use term in some capacity
+    (void)convert_mod(to_mod_expr(expr));
+
+    convert_expr(expr.op0());
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/smt2/smt2_conv.cpp:1263]: There shouldn't be any need for this, we should just print the term we get from `convert_mod`

File: src/solvers/smt2/smt2_conv.cpp [hunk 2]
```diff
@@ -2890,23 +2896,15 @@ void smt2_convt::convert_constant(const constant_exprt &expr)
...
-      out << "(bvsrem ";
+  PRECONDITION(
+    expr.type().id() == ID_unsignedbv || expr.type().id() == ID_signedbv);
 
-    convert_expr(expr.op0());
-    out << " ";
-    convert_expr(expr.op1());
-    out << ")";
-  }
+  if(expr.type().id() == ID_unsignedbv)
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/smt2/smt2_conv.cpp:2904]: The fact that this has operands that need to be `convert_expr`'d means it's not a leaf right?

File: src/solvers/smt2/smt2_conv.h [hunk 3]
```diff
@@ -31,6 +31,22 @@ class constant_exprt;
...
+  // Bitvector arithmetic
+  bvurem,
+  bvsrem,
+  bvudiv,
+  bvsdiv
+};
+
+struct SMT_Term
+{
+  SMT_Ops op;
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/smt2/smt2_conv.h:47]: This doesn't make a lot of sense ATM, I guess this is just a placeholder?

File: src/solvers/smt2/smt2_conv.h [hunk 4]
```diff
@@ -31,6 +31,22 @@ class constant_exprt;
 class index_exprt;
 class member_exprt;
 
+// TODO: This should be replaced on the spot with the
+//  smt-switch ops.h class
```
Comment (hannes-steffenhagen-diffblue) [inline src/solvers/smt2/smt2_conv.h:35]: Why though? We should just return SMT switch `Term`, which doesn't seem to reference ops? For what we're doing with these dummy terms I don't think we need anything more complicated than rose-tree of string, or even just string if it comes down to it.

**Comment by NlightNFotis** [comment]
There has been a change in priorities and this work is now parked.

**Pull request closed without merging**